lib/SILOptimizer/Utils/Differentiation/Common.cpp

```diff
@@ -23,5 +23,178 @@ namespace autodiff {
 
 raw_ostream &getADDebugStream() { return llvm::dbgs() << "[AD] "; }
 
+//===----------------------------------------------------------------------===//
+// Code emission utilities
+//===----------------------------------------------------------------------===//
+
+void collectAllActualResultsInTypeOrder(
+    ApplyInst *ai, ArrayRef<SILValue> extractedDirectResults,
+    SmallVectorImpl<SILValue> &results) {
+  auto calleeConvs = ai->getSubstCalleeConv();
+  unsigned indResIdx = 0, dirResIdx = 0;
+  for (auto &resInfo : calleeConvs.getResults()) {
+    results.push_back(resInfo.isFormalDirect()
+                          ? extractedDirectResults[dirResIdx++]
+                          : ai->getIndirectSILResults()[indResIdx++]);
+  }
+}
+
+SILValue joinElements(ArrayRef<SILValue> elements, SILBuilder &builder,
+                      SILLocation loc) {
+  if (elements.size() == 1)
+    return elements.front();
+  return builder.createTuple(loc, elements);
+}
+
+void extractAllElements(SILValue value, SILBuilder &builder,
```

Comment:
This is void SILBuilder::emitDestructureValueOperation(SILLocation loc, SILValue operand,
                                     SmallVectorImpl<SILValue> &result);

Reply:
Addressing in https://github.com/apple/swift/pull/30970#discussion_r407013548, let's continue discussion there.